File: src/web/static/js/stage4.js
const API_BASE = "/api";

const STAGE4_STATUS_LABELS = {
    url_ready: "Готово к сбору объявлений",
    listings_fresh: "Объявления уже собраны",
    listings_stale: "Нужно обновить объявления",
};

let stage4Offset = 0;
const stage4Limit = 20;
let stage4Total = 0;
let stage4Items = [];
let selectedItems = new Set();
let selectedRegNumber = null;

document.addEventListener("DOMContentLoaded", () => {
    loadStage4(true);
    document.getElementById("btn-run-stage4").addEventListener("click", runStage4);
    document.getElementById("btn-load-more-stage4").addEventListener("click", () => loadStage4(false));
    document.getElementById("select-all-stage4").addEventListener("change", (e) => toggleSelectAllVisible(e.target.checked));
});

function escapeHtml(value) {
    return String(value ?? "")
        .replaceAll("&", "&amp;")
        .replaceAll("<", "&lt;")
        .replaceAll(">", "&gt;")
        .replaceAll('"', "&quot;")
        .replaceAll("'", "&#39;");
}

function setStage4Status(message, type = "info") {
    const el = document.getElementById("stage4-status");
    if (!message) {
        el.style.display = "none";
        el.className = "status-box";
        el.textContent = "";
        return;
    }
    el.style.display = "block";
    el.className = `status-box ${type === "info" ? "" : type}`.trim();
    el.textContent = message;
}

function clearErrors() {
    const panel = document.getElementById("stage4-errors");
    panel.style.display = "none";
    panel.innerHTML = "";
}

function renderErrors(errors) {
    const panel = document.getElementById("stage4-errors");
    if (!errors || errors.length === 0) {
        clearErrors();
        return;
    }

    const rows = errors.map((line) => `<div class="error-row">${escapeHtml(line)}</div>`).join("");
    panel.innerHTML = `<strong>Ошибки запуска Этапа 4</strong>${rows}`;
    panel.style.display = "block";
}

async function loadStage4(reset = false, keepSelected = false) {
    try {
        if (reset) {
            stage4Offset = 0;
            stage4Items = [];
            selectedItems.clear();
            if (!keepSelected) selectedRegNumber = null;

            document.getElementById("stage4-list").innerHTML = "";
            document.getElementById("stage4-workspace").innerHTML = '<div class="empty-state">Выберите закупку слева</div>';
            if (!keepSelected) {
                clearErrors();
                setStage4Status("");
            }
        }

        const resp = await fetch(`${API_BASE}/stage4?offset=${stage4Offset}&limit=${stage4Limit}`);
        if (!resp.ok) throw new Error("Не удалось загрузить список закупок Этапа 4");

        const data = await resp.json();
        const items = Array.isArray(data) ? data : data.items || [];
        stage4Total = Array.isArray(data) ? items.length : data.total || 0;

        stage4Items = stage4Items.concat(items);
        renderList(items, reset);
        updateRunButton();

        stage4Offset += items.length;
        const btnLoadMore = document.getElementById("btn-load-more-stage4");
        if (!Array.isArray(data) && stage4Offset < stage4Total) {
            btnLoadMore.style.display = "inline-block";
            btnLoadMore.disabled = false;
        } else {
            btnLoadMore.style.display = "none";
        }

        if (reset && keepSelected && selectedRegNumber) {
            const exists = stage4Items.some((x) => x.reg_number === selectedRegNumber);
            if (exists) await selectZakupka(selectedRegNumber);
        }
    } catch (e) {
        console.error(e);
        setStage4Status(e.message || "Ошибка загрузки Этапа 4", "error");
    }
}

function renderList(items, reset) {
    const list = document.getElementById("stage4-list");
    if (reset) list.innerHTML = "";

    if (reset && items.length === 0) {
        list.innerHTML = '<div style="padding:20px; text-align:center; color:#999">Нет закупок для Этапа 4</div>';
        return;
    }

    items.forEach((item) => {
        const checked = selectedItems.has(item.reg_number);
        const hasStage3Url = !!item.two_gis_url;
        const statusLabel = STAGE4_STATUS_LABELS[item.status] || item.status || "—";

        const div = document.createElement("div");
        div.className = "list-item";
        div.dataset.reg = item.reg_number;

        div.innerHTML = `
            <input type="checkbox" class="row-select-stage4" ${checked ? "checked" : ""} ${hasStage3Url ? "" : "disabled"} onclick="event.stopPropagation(); toggleSelection('${item.reg_number}', this.checked)">
            <div class="list-item-content">
                <div class="list-item-header">${item.reg_number}</div>
                <div class="list-item-desc">${item.ai_city || "—"} | ${item.bid_end_date || "—"} | ${formatPrice(item.initial_price)}</div>
                <div class="list-item-desc">${statusLabel}</div>
                <div class="list-item-desc" style="color:${hasStage3Url ? "#64748b" : "#b91c1c"};">${hasStage3Url ? "Ссылка Этапа 3 есть" : "Ссылки нет (Этап 3)"}</div>
            </div>
        `;

        div.addEventListener("click", () => selectZakupka(item.reg_number));
        list.appendChild(div);
    });
}

function formatPrice(price) {
    if (!price) return "—";
    return `${(price / 1000000).toFixed(2)} млн`;
}

function toggleSelection(regNumber, checked) {
    if (checked) selectedItems.add(regNumber);
    else selectedItems.delete(regNumber);
    updateRunButton();
}

function toggleSelectAllVisible(checked) {
    const checkboxes = document.querySelectorAll("#stage4-list .row-select-stage4");
    checkboxes.forEach((cb) => {
        if (cb.disabled) return;
        cb.checked = checked;

        const reg = cb.closest(".list-item")?.dataset?.reg;
        if (!reg) return;
        if (checked) selectedItems.add(reg);
        else selectedItems.delete(reg);
    });
    updateRunButton();
}

function updateRunButton() {
    const btn = document.getElementById("btn-run-stage4");
    btn.disabled = selectedItems.size === 0;
    btn.textContent = selectedItems.size > 0
        ? `Собрать объявления (${selectedItems.size})`
        : "Собрать объявления";
}

function renderSourceLinks(item) {
    let external = item.external_source || "";
    const externalUrl = item.external_url || "";

    if (!external && externalUrl) {
        const lowered = externalUrl.toLowerCase();
        if (lowered.includes("domclick") || lowered.includes("dom.click")) external = "domclick";
        else if (lowered.includes("cian")) external = "cian";
        else if (lowered.includes("avito")) external = "avito";
    }

    const domclick = external === "domclick" && externalUrl ? `<a class="source-link" href="${externalUrl}" target="_blank">Домклик</a>` : "—";
    const cian = external === "cian" && externalUrl ? `<a class="source-link" href="${externalUrl}" target="_blank">Циан</a>` : "—";
    const avito = external === "avito" && externalUrl ? `<a class="source-link" href="${externalUrl}" target="_blank">Avito</a>` : "—";
    const twoGisCard = item.two_gis_url ? `<a class="source-link" href="${item.two_gis_url}" target="_blank">2ГИС</a>` : "—";

    return `<div class="links-row"><strong>Ссылки:</strong> ${twoGisCard} | ${domclick} | ${cian} | ${avito}</div>`;
}

async function selectZakupka(regNumber) {
    selectedRegNumber = regNumber;
    document.querySelectorAll(".list-item").forEach((el) => el.classList.remove("active"));
    document.querySelector(`.list-item[data-reg="${regNumber}"]`)?.classList.add("active");

    const workspace = document.getElementById("stage4-workspace");
    workspace.innerHTML = '<div class="empty-state">Загрузка объявлений...</div>';
    const selectedMeta = stage4Items.find((x) => x.reg_number === regNumber);

    if (!selectedMeta || !selectedMeta.two_gis_url) {
        workspace.innerHTML = `
            <div class="workspace-header">
                <div><strong>${regNumber}</strong></div>
                <div style="color:#b91c1c;">Ссылки нет (Этап 3)</div>
            </div>
            <div class="workspace-body">
                <div class="listing-card" style="border-color:#fecaca; background:#fff1f2;">
                    Для этой закупки нет ссылки Этапа 3. Сначала сформируйте ссылку на Этапе 3.
                </div>
            </div>
        `;
        return;
    }

    try {
        const resp = await fetch(`${API_BASE}/stage4/${regNumber}/listings`);
        if (!resp.ok) throw new Error("Не удалось загрузить объявления");

        const data = await resp.json();
        const items = data.items || [];

        const head = `
            <div class="workspace-header">
                <div><strong>${regNumber}</strong></div>
                <div>Собрано объявлений: <strong>${items.length}</strong></div>
                <div><a href="${selectedMeta.two_gis_url}" target="_blank">Ссылка Этапа 3</a></div>
            </div>
        `;

        if (items.length === 0) {
            workspace.innerHTML = `${head}<div class="workspace-body"><div class="empty-state">По этой закупке пока нет собранных объявлений</div></div>`;
            return;
        }

        const cards = items.map((item) => {
            const rooms = item.rooms != null ? `${item.rooms} комн.` : "—";
            const area = item.area_m2 != null ? `${item.area_m2} м²` : "—";
            const floor = item.floor != null ? `${item.floor}/${item.building_floors || "?"}` : "—";
            const year = item.building_year != null ? `${item.building_year}` : "—";
            const price = item.price_rub ? `${item.price_rub.toLocaleString("ru-RU")} ₽` : "—";

            const links = renderSourceLinks(item);

            return `
                <div class="listing-card">
                    <div class="listing-title">#${item.rank || "—"} — ${price}</div>
                    <div class="listing-meta">
                        <div><strong>Адрес:</strong> ${item.address || "—"}</div>
                        <div><strong>Характеристики:</strong> ${rooms}, ${area}, этаж ${floor}, год ${year}</div>
                        ${links}
                    </div>
                </div>
            `;
        }).join("");

        workspace.innerHTML = `${head}<div class="workspace-body">${cards}</div>`;
    } catch (e) {
        console.error(e);
        workspace.innerHTML = '<div class="empty-state">Не удалось загрузить объявления</div>';
    }
}

function extractApiError(payload) {
    if (!payload) return "";
    if (typeof payload.detail === "string") return payload.detail;
    if (Array.isArray(payload.detail)) {
        return payload.detail.map((x) => x.msg || JSON.stringify(x)).join("; ");
    }
    return "";
}

async function runStage4() {
    if (selectedItems.size === 0) {
        setStage4Status("Выберите закупки галочками для запуска.", "error");
        return;
    }

    const topN = Number(document.getElementById("top-n-stage4").value || 20);
    const getDetails = !!document.getElementById("details-stage4").checked;
    const btn = document.getElementById("btn-run-stage4");
    const prevSelectedReg = selectedRegNumber;

    btn.disabled = true;
    setStage4Status("Собираю объявления по выбранным закупкам...");
    clearErrors();

    try {
        const resp = await fetch(`${API_BASE}/actions/run_stage4`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                reg_numbers: Array.from(selectedItems),
                top_n: topN,
                get_details: getDetails,
            }),
        });

        const result = await resp.json();
        if (!resp.ok) {
            const msg = extractApiError(result) || `HTTP ${resp.status}`;
            throw new Error(msg);
        }

        setStage4Status(`Готово: обработано закупок ${result.processed || 0}, собрано объявлений ${result.total_listings || 0}.`, "success");
        renderErrors(result.errors || []);

        selectedRegNumber = prevSelectedReg;
        await loadStage4(true, true);
    } catch (e) {
        console.error(e);
        setStage4Status(`Ошибка запуска: ${e.message || e}`, "error");
        renderErrors([`Ошибка запуска: ${e.message || e}`]);
    } finally {
        updateRunButton();
    }
}

File: src/api/routes/stage4.py
"""Stage 4 routes."""

from fastapi import APIRouter, Depends

from models.statuses import STAGE4_PROCESSED_STATUSES, STAGE4_QUEUE_STATUSES, ZakupkaStatus

from .common import RunStage4Request, admin_required


router = APIRouter()


@router.get("/api/stage4")
def get_stage4_data(
    offset: int = 0,
    limit: int = 20,
    view: str = "queue",
    admin: bool = Depends(admin_required),
):
    from ..app import get_pipeline

    pipeline = get_pipeline()
    stage4_statuses = STAGE4_PROCESSED_STATUSES if view == "processed" else STAGE4_QUEUE_STATUSES
    rows, total = pipeline.db.zakupki.get_stage4_page_with_ai_city(
        statuses=list(stage4_statuses),
        offset=offset,
        limit=limit,
        processed_view=(view == "processed"),
    )
    items = []
    for row in rows:
        items.append(
            {
                "reg_number": row.get("reg_number", ""),
                "description": row.get("description"),
                "update_date": row.get("update_date"),
                "bid_end_date": row.get("bid_end_date"),
                "initial_price": row.get("initial_price"),
                "link": row.get("link"),
                "two_gis_url": row.get("two_gis_url"),
                "status": row.get("status"),
                "ai_city": row.get("ai_city"),
            }
        )
    return {"items": items, "total": total, "offset": offset, "limit": limit, "view": view}


@router.get("/api/stage4/{reg_number}/listings")
def get_stage4_listings(reg_number: str, admin: bool = Depends(admin_required)):
    from ..app import get_pipeline

    pipeline = get_pipeline()
    listings = pipeline.db.listings.get_for_zakupka(reg_number)
    items = []
    for i, listing in enumerate(listings, 1):
        d = listing.to_dict() if hasattr(listing, "to_dict") else dict(listing)
        items.append(
            {
                "rank": d.get("rank") if d.get("rank") is not None else i,
                "price_rub": d.get("price_rub"),
                "address": d.get("address"),
                "rooms": d.get("rooms"),
                "area_m2": d.get("area_m2"),
                "floor": d.get("floor"),
                "building_floors": d.get("building_floors"),
                "building_year": d.get("building_year"),
                "two_gis_url": d.get("two_gis_url"),
                "external_source": d.get("external_source"),
                "external_url": d.get("external_url"),
            }
        )
    return {"reg_number": reg_number, "items": items, "total": len(items)}


@router.post("/api/actions/run_stage4")
def run_stage4(req: RunStage4Request, admin: bool = Depends(admin_required)):
    from ..app import get_pipeline

    pipeline = get_pipeline()
    if not req.reg_numbers:
        return {"status": "warning", "message": "reg_numbers is required"}

    zakupki_by_reg = {z.reg_number: z for z in pipeline.db.zakupki.get_by_reg_numbers(req.reg_numbers)}
    processed = 0
    total_listings = 0
    errors = []

    for reg_number in req.reg_numbers:
        z = zakupki_by_reg.get(reg_number)
        if not z:
            errors.append(f"{reg_number}: закупка не найдена")
            continue
        if z.status not in STAGE4_QUEUE_STATUSES:
            errors.append(
                f"{reg_number}: статус '{z.status}', ожидается '{ZakupkaStatus.URL_READY}' или '{ZakupkaStatus.STAGE4_ERROR}'"
            )
            continue
        if not z.two_gis_url:
            errors.append(f"{reg_number}: Ссылки нет (Этап 3)")
            continue
        try:
            result = pipeline.run_stage4_for_zakupka(
                reg_number=z.reg_number,
                url=z.two_gis_url,
                top_n=req.top_n,
                get_details=req.get_details,
            )
            if result.error:
                errors.append(f"{z.reg_number}: {result.error}")
            else:
                processed += 1
                total_listings += result.actual_n
            pipeline.logger.info(
                "stage_progress reg_number=%s stage=4 result=%s reason=%s",
                z.reg_number,
                "error" if result.error else "ok",
                result.error or f"listings={result.actual_n}",
            )
        except Exception as e:
            errors.append(f"{z.reg_number}: {e}")
            pipeline.db.zakupki.update_status(z.reg_number, ZakupkaStatus.STAGE4_ERROR)
            pipeline.logger.info(
                "stage_progress reg_number=%s stage=4 result=error reason=%s",
                z.reg_number,
                e,
            )

    return {
        "status": "ok" if processed > 0 and not errors else ("warning" if processed > 0 else "error"),
        "message": f"Собрано {total_listings} объявлений из {processed} закупок",
        "processed": processed,
        "total_listings": total_listings,
        "errors": errors,
    }
